core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java

```diff
@@ -0,0 +1,56 @@
+package org.testcontainers.utility;
+
+import com.github.dockerjava.api.model.Container;
+import com.github.dockerjava.api.model.PruneType;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link ResourceReaper} implementation that uses {@link Runtime#addShutdownHook(Thread)}
+ * to cleanup containers.
+ */
+class JVMHookResourceReaper extends ResourceReaper {
+
+    JVMHookResourceReaper() {
+        setHook();
+    }
+
+    @Override
+    public synchronized void performCleanup() {
+        super.performCleanup();
```

Comment:
Why do we first have to call the super method (which will call)
```java
  registeredContainers.forEach(this::removeContainer);
  registeredNetworks.forEach(this::removeNetwork);
  registeredImages.forEach(this::removeImage);
```

and then call prune afterward?

Reply:
`forEach` methods will cleanup resources explicitly registered with `registerContainerForCleanup`, while `DEATH_NOTE`-based ones will be cleaning by labels

Reply:
But the set of resources each approach would remove, is very likely to overlap, or?

Reply:
it is actually the other way around - very unlikely.

We don't register containers/networks/images by hands and rely on the filter-based approach. I deprecated the methods, so we can remove them in the future, but for now we have to keep both.

Reply:
That makes much more sense then, thanks for the explanation.